### PR TITLE
Inspect command for mgadm

### PR DESCRIPTION
**Pull request by mbussolotto**
Inspect command take an image as argument and return information from image and deployment (if present). The function created will be used internally for gathering information for upgrade

### CODE REVIEW

File: mgradm/cmd/inspect/shared/shared.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+func GenerateInspectScript(scriptDir string) {
+
+	data := templates.InspectTemplateData{
+		Param:      inspectValues,
+		OutputFile: InspectOutputFile.Directory + "/" + InspectOutputFile.Basename,
+	}
+
+	scriptPath := filepath.Join(scriptDir, InspectScriptFilename)
+	if err := utils.WriteTemplateToFile(data, scriptPath, 0555, true); err != nil {
+		log.Fatal().Err(err).Msgf("Failed to generate inspect script")
```
Comment (cbosdo) [inline mgradm/cmd/inspect/shared/shared.go:49]:
```suggestion
		return fmt.Errorf("Failed to generate inspect script: %s", err)
```

File: mgradm/cmd/inspect/shared/shared.go [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+	}
```
Comment (cbosdo) [inline mgradm/cmd/inspect/shared/shared.go:50]:
```suggestion
	}
	return nil
```

File: mgradm/cmd/inspect/shared/shared.go [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+}
+
+func ReadInspectData(scriptDir string) map[string]string {
```
Comment (cbosdo) [inline mgradm/cmd/inspect/shared/shared.go:53]:
```suggestion
func ReadInspectData(scriptDir string) (map[string]string, error) {
```

File: mgradm/cmd/inspect/shared/shared.go [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+	path := filepath.Join(scriptDir, "data")
+
+	log.Debug().Msgf("Trying to read %s", path)
+
+	data, err := os.ReadFile(path)
+
+	if err != nil {
+		log.Fatal().Err(err).Msgf("Cannot parse file")
```
Comment (cbosdo) [inline mgradm/cmd/inspect/shared/shared.go:61]:
```suggestion
		return [string]string{}, fmt.Errorf("Cannot parse file %s: %s", path, err)
```

File: mgradm/cmd/inspect/shared/shared.go [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+	viper.SetConfigType("env")
+	viper.ReadConfig(bytes.NewBuffer(data))
+
+	for _, v := range inspectValues {
+		if len(viper.GetString(v.Variable)) > 0 {
+			inspectResult[v.Variable] = viper.GetString(v.Variable)
+		}
+	}
+	return inspectResult
```
Comment (cbosdo) [inline mgradm/cmd/inspect/shared/shared.go:74]:
```suggestion
	return inspectResult, nil
```

File: mgradm/shared/podman/podman.go [hunk 6]
```diff
@@ -119,3 +119,22 @@ func UpdateSslCertificate(cnx *shared.Connection, chain *ssl.CaChain, serverPair
 	log.Info().Msg("Restarting services after updating the certificate")
 	utils.RunCmdStdMapping("podman", "exec", podman.ServerContainerName, "spacewalk-service", "restart")
 }
+
+func RunContainer(name string, image string, extraArgs []string, cmd []string) {
```
Comment (cbosdo) [inline mgradm/shared/podman/podman.go:123]:
```suggestion
func RunContainer(name string, image string, extraArgs []string, cmd []string) error {
```

File: mgradm/shared/podman/podman.go [hunk 7]
```diff
@@ -119,3 +119,22 @@ func UpdateSslCertificate(cnx *shared.Connection, chain *ssl.CaChain, serverPair
...
+		podmanArgs = append(podmanArgs, "-v", volumeName+":"+containerPath)
+	}
+
+	podmanArgs = append(podmanArgs, image)
+	podmanArgs = append(podmanArgs, cmd...)
+
+	err := utils.RunCmdStdMapping("podman", podmanArgs...)
+
+	if err != nil {
+		log.Fatal().Err(err).Msgf("Failed to run %s container", name)
```
Comment (cbosdo) [inline mgradm/shared/podman/podman.go:138]:
```suggestion
		return fmt.Errorf("Failed to run %s container: %s", name, err)
```

File: mgradm/shared/podman/podman.go [hunk 8]
```diff
@@ -119,3 +119,22 @@ func UpdateSslCertificate(cnx *shared.Connection, chain *ssl.CaChain, serverPair
...
+	}
```
Comment (cbosdo) [inline mgradm/shared/podman/podman.go:139]:
```suggestion
	}
	return nil
```

File: shared/kubernetes/utils.go [hunk 9]
```diff
@@ -128,3 +128,80 @@ func GetPullPolicy(name string) string {
...
+	out, err := utils.RunCmdOutput(zerolog.TraceLevel, "kubectl", args...)
+	if err != nil {
+		if replicas, _ := strconv.Atoi(string(out)); replicas > 0 {
+			return true
+		}
+	}
+	return false
+}
+
+func RunPod(podname string, image string, pullPolicy string, command string, args ...string) {
```
Comment (cbosdo) [inline shared/kubernetes/utils.go:146]:
```suggestion
func RunPod(podname string, image string, pullPolicy string, command string, args ...string) error {
```

File: shared/kubernetes/utils.go [hunk 10]
```diff
@@ -128,3 +128,80 @@ func GetPullPolicy(name string) string {
...
+	arguments := []string{"run", podname, "--image", image, "--image-pull-policy", pullPolicy}
+
+	arguments = append(arguments, args...)
+	arguments = append(arguments, "--command", "--", command)
+	err := utils.RunCmdStdMapping("kubectl", arguments...)
+
+	if err != nil {
+		log.Fatal().Err(err).Msgf("Cannot run %s using image %s", command, image)
```
Comment (cbosdo) [inline shared/kubernetes/utils.go:154]:
```suggestion
		return fmt.Errorf("Cannot run %s using image %s: %s", command, image, err)
```

File: shared/kubernetes/utils.go [hunk 11]
```diff
@@ -128,3 +128,80 @@ func GetPullPolicy(name string) string {
...
+	}
```
Comment (cbosdo) [inline shared/kubernetes/utils.go:155]:
```suggestion
	}
	return nil
```

File: shared/kubernetes/utils.go [hunk 12]
```diff
@@ -128,3 +128,80 @@ func GetPullPolicy(name string) string {
...
+}
+
+func DeletePod(podname string) string {
```
Comment (cbosdo) [inline shared/kubernetes/utils.go:158]:
```suggestion
func DeletePod(podname string) (string, error) {
```

File: shared/kubernetes/utils.go [hunk 13]
```diff
@@ -128,3 +128,80 @@ func GetPullPolicy(name string) string {
...
+	arguments := []string{"delete", "pod", podname}
+
+	out, err := utils.RunCmdOutput(zerolog.DebugLevel, "kubectl", arguments...)
+
+	if err != nil {
+		log.Fatal().Err(err).Msgf("Cannot %s", arguments)
```
Comment (cbosdo) [inline shared/kubernetes/utils.go:164]:
```suggestion
		return nil, fmt.Errorf("Cannot %s: %s", strings.Join(" ", arguments), err)
```

File: shared/kubernetes/utils.go [hunk 14]
```diff
@@ -128,3 +128,80 @@ func GetPullPolicy(name string) string {
...
+	}
+	return string(out)
```
Comment (cbosdo) [inline shared/kubernetes/utils.go:166]:
```suggestion
	return string(out), nil
```

File: shared/kubernetes/utils.go [hunk 15]
```diff
@@ -128,3 +128,80 @@ func GetPullPolicy(name string) string {
...
+}
+
+func WaitForPod(podname string, status string) {
```
Comment (cbosdo) [inline shared/kubernetes/utils.go:169]:
```suggestion
func WaitForPod(podname string, status string) error {
```

File: shared/kubernetes/utils.go [hunk 16]
```diff
@@ -128,3 +128,80 @@ func GetPullPolicy(name string) string {
...
+
+		if strings.ToUpper(string(out)) == strings.ToUpper(status+"\n") {
+			log.Debug().Msgf("%s pod status is %s", podname, status)
+			break
+		}
+		if err != nil {
+			log.Fatal().Err(err).Msgf("Pod %s does not %s in %s seconds", podname, out, strconv.Itoa(waitSeconds))
+		}
+		time.Sleep(1 * time.Second)
+	}
```
Comment (cbosdo) [inline shared/kubernetes/utils.go:188]:
```suggestion
    var err error
	for i := 0; i < waitSeconds; i++ {
		out, err := utils.RunCmdOutput(zerolog.DebugLevel, "kubectl", cmdArgs...)
		log.Debug().Msgf("%s status is %s", podname, out)

		if strings.ToUpper(string(out)) == strings.ToUpper(status+"\n") {
			log.Debug().Msgf("%s pod status is %s", podname, status)
			break
		}
		time.Sleep(1 * time.Second)
	}
	if err != nil {
		return fmt.Errorf("Pod %s does not %s in %s seconds: %s", podname, out, strconv.Itoa(waitSeconds), err)
	}
	return nil
```

File: mgradm/cmd/inspect/inspect.go [hunk 17]
```diff
@@ -0,0 +1,29 @@
...
+	"github.com/uyuni-project/uyuni-tools/mgradm/cmd/inspect/podman"
+	"github.com/uyuni-project/uyuni-tools/shared/types"
+)
+
+func NewCommand(globalFlags *types.GlobalFlags) *cobra.Command {
+
+	inspectCmd := &cobra.Command{
+		Use:   "inspect",
+		Short: "inspect",
+		Long:  "Extract information from image or deployment",
```
[the remark this answers is not in the document]
Comment (mbussolotto) [reply mgradm/cmd/inspect/inspect.go:19]: correct

**Comment by mbussolotto** [comment]
@cbosdo , thanks for all your comment, now should be fine. Changes are:

- error handling
- remove `podman`/`kubernetes` argument
- improve argument parameter: if image and or tag are not passed, the value are detected by the running image

**Comment by mbussolotto** [comment]
Closing this PR, this change would be implemented in upgrade PR

**Pull request closed without merging**